### DISCUSSION TITLE
IMAPIdleExample.cs/ProtocolLogger Problem

### DISCUSSION THREAD

**Discussion by Latinib** [jstedfast/MailKit#1664] [Q&A]
I'm able to run the  IMAPIdleExample.cs project downloaded from github as a console app and under a service background task. However, I'm not able to run it as part of a Windows form app.

The problem seems related to the ProtocolLogger. When I pass Console.StandardError to its constructor, I get an exception "The CancellationTokenSrource has been disposed" after client.AuthenticateAsync in ReconnectAsync.

When I pass a the name of a local file to the ProtocolLogger constructor, I get an exception "Cannot access a closed file" after client.ConnectAsync in ReconnectAsync. Note that I also use a local file in the console and service versions of the app and have no problems.

Environment: Windows 11 Pro, Visual Studio 2022 64-bit, .Net 6

Any advice would be greatly appreciated.

**Comment by jstedfast** [comment]
In both cases it sounds like you've disposed the ProtocolLogger or the file stream or something.

You'll have to paste your code.

**Reply by Latinib** [reply]
Respectfully, I don't see how I could be disposing the ProtocolLogger or file stream. All I do is pass a file name to ProtocolLogger via the constructor and invoke RunAsync. After that, it's all under IdleClient's control.

Please see the attached VS 2022 solution. It consists of three projects:

IdleClient: the IdleClient class embedded in a class library. It's the original IdleClient from IdleClientExample with minimal modifications to allow it to be invoked from either a console app or a WinForms app and try/catch code in ReconnectAsync to report exceptions. It includes an additional constructor argument to allow Protocol Logger to output to a local file instead of Console.OpenStandardError and another constructor argument to allow Console.WriteLine outputs to be displayed in a textbox instead of Console

ConsoleApp1: program.cs from IdleClientExample modified to invoke IdleClientLib.IdleClient instead of the embedded class and direct ProtocolLogger outputs to a local file

WinFormsApp1: a simple Windows form project to invoke IdleClientLib.IdleClient, direct protocol outputs to a local file and display other Console outputs in a textbox

ConsoleApp1 works fine but WinFormsApp1 gets exceptions in ReconnectAsync.

[Mailkit.zip](https://github.com/jstedfast/MailKit/files/13314036/Mailkit.zip)

**Reply by jstedfast** [reply]
I happen to be correct :-)

In your button handler, you instantiate a new IdleClient in a `using` statement.

What happens at the end of a `using` block? The client object gets disposed, correct?

Inside the using block, you call:

```csharp
var idleTask = client.RunAsync();
```

But you aren't "awaiting" it.

client.RunAsync() just queues asynchronous operations to run sometime down the road. It doesn't execute the content of the RunAsync() method *now* in a synchronous manner.

This means that the end of the using block is hit by the runtime *before* the content of the RunAsync() method are executed. Which means the IdleClient gets disposed. Which means the ProtocolLogger gets disposed and the CancellationTokenSource, etc all get disposed as well.

What is happening is that the btnIdle_Click() method exits execution before the IdleClient.RunAsync() content even starts running (due to the way async methods work) and so, when the IdleCLient.RunAsync() code finally does execute, everything is already disposed and things crash.

Trust me, I've been bitten by this type of bug before so I knew it when I saw it. It was the first thing I looked for in your Forms app.

**Reply by Latinib** [reply]
Thanks for the explanation.
